ydb/src/client_topic/topicwriter/writer.rs

```diff
@@ -0,0 +1,405 @@
+use crate::client_topic::topicwriter::writer_reception_queue::{
+    TopicWriterReceptionQueue, TopicWriterReceptionTicket, TopicWriterReceptionType,
+};
+use crate::grpc_connection_manager::GrpcConnectionManager;
+use crate::grpc_wrapper::raw_errors::{RawError, RawResult};
+use crate::grpc_wrapper::raw_topic_service::common::codecs::RawSupportedCodecs;
+use crate::{grpc_wrapper, YdbError, YdbResult};
+use futures_util::{StreamExt, TryStreamExt};
+use std::borrow::{Borrow, BorrowMut};
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::future::Future;
+use std::ops::Deref;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex, MutexGuard};
+use std::task::{Context, Poll};
+use std::time::Instant;
+use std::time::{Duration, UNIX_EPOCH};
+use tokio::select;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::Receiver;
+use tokio::task::JoinHandle;
+use tokio::time::error::Elapsed;
+use tokio::time::timeout;
+use tokio_util::sync::CancellationToken;
+use ydb_grpc::ydb_proto::topic::stream_write_message;
+use ydb_grpc::ydb_proto::topic::stream_write_message::from_client::ClientMessage;
+use ydb_grpc::ydb_proto::topic::stream_write_message::from_server::ServerMessage;
+use ydb_grpc::ydb_proto::topic::stream_write_message::init_request::Partitioning;
+use ydb_grpc::ydb_proto::topic::stream_write_message::write_request::{message_data, MessageData};
+use ydb_grpc::ydb_proto::topic::stream_write_message::{
+    FromClient, FromServer, InitRequest, WriteRequest,
+};
+
+pub enum TopicWriterState {
+    Working,
+    FinishedOk,
+    FinishedWithError(YdbError),
+}
+
+#[allow(dead_code)]
+pub struct TopicWriter {
+    pub(crate) path: String,
+    pub(crate) producer_id: Option<String>,
+    pub(crate) partition_id: i64,
+    pub(crate) session_id: String,
+    pub(crate) last_seq_num_handled: i64,
+    pub(crate) write_request_messages_chunk_size: usize,
+    pub(crate) write_request_send_messages_period: Duration,
+
+    pub(crate) auto_set_seq_no: bool,
+    pub(crate) codecs_from_server: RawSupportedCodecs,
+
+    writer_message_sender: mpsc::Sender<TopicWriterMessage>,
+    writer_loop: JoinHandle<()>,
+    receive_messages_loop: JoinHandle<()>,
+
+    cancellation_token: CancellationToken,
+    writer_state: Arc<Mutex<TopicWriterState>>,
+
+    confirmation_reception_queue: Arc<Mutex<TopicWriterReceptionQueue>>,
+
+    pub(crate) connection_manager: GrpcConnectionManager,
+}
+
+#[allow(dead_code)]
+pub struct AckFuture {}
+
+impl Future for AckFuture {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        unimplemented!("prototype")
+    }
+}
+
+struct WriterPeriodicTaskParams {
+    write_request_messages_chunk_size: usize,
+    write_request_send_messages_period: Duration,
+    auto_set_seq_no: bool,
+    producer_id: Option<String>,
+    request_stream: mpsc::UnboundedSender<stream_write_message::FromClient>,
+}
+
+impl TopicWriter {
+    pub(crate) async fn new(
```

Comment:
A long function difficult to read and test.

I suggest to split the function for smaller functions:
1. Create internal state
2. Reconnect loop
3. Initialize stream (connect to server, handshake by init request/response)
4. Loop func for send messages
5. Loop func for receive acks